core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala

```diff
@@ -0,0 +1,212 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.executor
+
+import org.apache.spark.{SparkConf, Logging, SecurityManager}
+import org.apache.mesos.{Executor => MesosExecutor, ExecutorDriver, MesosExecutorDriver, MesosNativeLibrary}
+import org.apache.spark.util.{Utils, SignalLogger}
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.mesos.Protos._
+import org.apache.spark.deploy.worker.StandaloneWorkerShuffleService
+import scala.collection.JavaConversions._
+import scala.io.Source
+import java.io.{File, PrintWriter}
+
+/**
+ * The Coarse grained Mesos executor backend is responsible for launching the shuffle service
+ * and the CoarseGrainedExecutorBackend actor.
+ * This is assuming the scheduler detected that the shuffle service is enabled and launches
+ * this class instead of CoarseGrainedExecutorBackend directly.
+ */
+private[spark] class CoarseGrainedMesosExecutorBackend(val sparkConf: SparkConf)
+  extends MesosExecutor
+  with Logging {
+
+  private var shuffleService: StandaloneWorkerShuffleService = null
+  private var driver: ExecutorDriver = null
+  private var executorProc: Process = null
+  private var taskId: TaskID = null
+  @volatile var killed = false
+
+  override def registered(
+      driver: ExecutorDriver,
+      executorInfo: ExecutorInfo,
+      frameworkInfo: FrameworkInfo,
+      slaveInfo: SlaveInfo) {
+    this.driver = driver
+    logInfo("Coarse Grained Mesos Executor '" + executorInfo.getExecutorId.getValue +
+      "' is registered.")
+
+    if (shuffleService == null) {
+      sparkConf.set("spark.shuffle.service.enabled", "true")
+      shuffleService = new StandaloneWorkerShuffleService(sparkConf, new SecurityManager(sparkConf))
+      shuffleService.startIfEnabled()
+    }
+  }
+
+  override def launchTask(d: ExecutorDriver, taskInfo: TaskInfo) {
+    if (executorProc != null) {
+      logError("Received LaunchTask while executor is already running")
+      val status = TaskStatus.newBuilder()
+        .setTaskId(taskInfo.getTaskId)
+        .setSlaveId(taskInfo.getSlaveId)
+        .setState(TaskState.TASK_FAILED)
+        .setMessage("Received LaunchTask while executor is already running")
+        .build()
+      d.sendStatusUpdate(status)
+      return
+    }
+
+    killed = false
+
+    logInfo("Launching task id: " + taskInfo.getTaskId.getValue)
+
+    // We are launching the CoarseGrainedExecutorBackend via subprocess
+    // because the backend is designed to run in its own process.
+    // Since it's a shared class we are preserving the existing behavior
+    // and launching it as a subprocess here.
+    val command = "exec " + Utils.deserialize[String](taskInfo.getData().toByteArray)
```

Comment:
here `command` is a giant string instead of a list of command line arguments. Should we do something like `Utils.splitCommandString` here to preserve the command properly when we pass it to the `ProcessBuilder`?